src/pint/toa.py

```diff
@@ -2166,9 +2166,14 @@ def merge_TOAs(TOAs_list):
     # We do not ensure that the command list is flat
     nt.commands = [tt.commands for tt in TOAs_list]
     # Now do the actual table stacking
-    nt.table = table.vstack(
-        [tt.table for tt in TOAs_list], join_type="exact", metadata_conflicts="silent"
-    )
+    start_index = np.maximum.reduce(nt.table["index"]) + 1
+    tables = []
+    for tt in TOAs_list:
+        t = copy.deepcopy(tt.table)
+        t["index"] += start_index
+        start_index += np.maximum.reduce(tt.table["index"]) + 1
+        tables.append(t)
+    nt.table = table.vstack(tables, join_type="exact", metadata_conflicts="silent")
```

Comment:
Actually, unless there is a good reason *not* to do it that way, I'd much prefer it as 1) it is a one-liner and 2) it is way more obvious what is going on and why (and it also has the benefit that the indices are by definition unique).

Reply:
If the reason is to keep the ordering of the indices that include the grouping by `"obs"`, that can be easily handled I think by doing a `tt.table.sort('index')` prior to the list comprehension that built the table.  The final `group_by` then gives you what you want.

Reply:
Because the separate TOAs objects might not have contiguous index arrays - someone might have taken a subset of the TOAs for one of them and want to be able to identify which TOAs in the original corresponded to that. I realize without a `max_index` member we can't quite achieve that, but I'd like to aim for that as a goal. The workflow I would like to enable is:

- Load in some TOAs objects `fulls` and select subsets of them `exciseds`.
- Merge `fulls` into `full` and `exciseds` into `excised`.
- Propagate flags from `excised` back to `full`.

Another reason to preserve the index entries is that they may well not be in order, and a major goal of them is that they record the original order of the TOAs (we could have `write_tim_file` emit them in index order if we wanted); one could argsort the tables in a for loop then put in an arange but that is even more work than the current design.

Reply:
Fair enough.  I do think this brings up the question as to what `index` really means, though.  Originally it literally meant the TOA number in the TOA table.  Currently, with a merge, it doesn't mean that anymore.  But maybe that was a bug?  Your workflow definitely breaks that definition of `index`.

Reply:
Your workflow doesn't work unless you do all of the excision on a merged `fulls`, right?  Because otherwise you lose the information about which numbers and how many TOAs were cut unless you reverse your index code (and keep a `max_index` for each `full`, as you mention)?  And on top of that, ` index` is *definitely* not meaning a table index anymore -- it is more just a "unique identifier".  I think I'd prefer that we fix the merge bug and that `index` really remains a table index (which, is obviously also a unique identifier).

Reply:
The current PR exists because @swiggumj wants to be able to identify individual TOA's in excised TOAs objects with the same TOA's in non-excised TOAs objects, so a table index completely fails to meet our users' needs.

Reply:
If we had the `max_index` member, then merging the excised TOAs objects results in the same numeration as merging the non-excised TOAs objects, and we keep the ability to identify TOAs across the merge. Without the `max_index` feature, no, that doesn't work because the last TOA might get excised. Hence wanting to add the `max_index`. But @swiggumj is trying to get automated outlier excision working and that is waiting on a fix very like this. So I want this to go in, solve his problem, and then I can introduce and test `max_index`.

Reply:
That's fine.  But that doesn't mean we should break what `index` is (or was) supposed to mean.  There are *many* simple ways to make this work simply and cleanly.  One really simple way is to make separate column that has unique and unchanging IDs for the TOAs and use that to figure out which TOAs are the ones in the original.  Another way is as you and I have mentioned for a while -- keep an array of your mask around and not a separate chopped TOAs object.

I'd prefer that we fixed the `index` bug in merging and have that remain to mean "location of the TOA in the original input" as is in our Docs now.  Every time we create a new TOAs object, `index` would therefore mean the same thing.

Reply:
To be clear, here is my proposal which fixes the `index` bug in `merge_TOAs` and gives you unique TOA ids that are in the same order as the original TOAs.  I think this allows @swiggumj to do what he wants (as long as he merges everything first) and it keeps the behavior of `index` consistent for TOAs objects created by PINT.
```
diff --git a/src/pint/toa.py b/src/pint/toa.py
index 26f735c5..07713eb4 100644
--- a/src/pint/toa.py
+++ b/src/pint/toa.py
@@ -2165,10 +2165,15 @@ def merge_TOAs(TOAs_list):
             nt.filename.append(xx)
     # We do not ensure that the command list is flat
     nt.commands = [tt.commands for tt in TOAs_list]
+    # Keep the oritinal TOAs objects sorted as they originall were
+    for tt in TOAs_list:
+        tt.table.sort("index")
     # Now do the actual table stacking
     nt.table = table.vstack(
         [tt.table for tt in TOAs_list], join_type="exact", metadata_conflicts="silent"
     )
+    # Reset the "index" column
+    nt.table["index"] = np.arange(nt.ntoas)
     # Fix the table meta data about filenames
     nt.table.meta["filename"] = nt.filename
     nt.hashes = {}
```

Reply:
I think my issue might be a bit mis-represented here. I haven't split excised TOAs into their own objects since it's easier for the time being to simply work with a table copy in the same object. Having unique indices is adequate for my purposes here. That way remaining indices in `toas.table` will map uniquely to those in `toas.orig_table` after making various cuts, which currently is not working as expected.

Reply:
I'm not sure I understand your solution for uniquely identifying TOAs? UUIDs are a bit cumbersome? Numbering them as they come out of the file accomplishes this within TOAs objects, and given we unpickle TOAs objects from previous runs we can't rely on some sort of global counter. We can't use a hash because people modify the individual TOAs. We could use tuples (file_number, toa_number) but Astropy doesn't have tuple columns as far as I know, and that complicates searching and sorting.

`index` only had a clear meaning for single-file TOAs objects, the introduction of merging basically made them unclear. The only use I could see for index tracking position in the original file was that we could restore the original order - which ability this preserves, if we treat the files as being merged in the order that TOAs objects are merged. This already happens when the merging goes through INCLUDEs, now `merge_toas` behaves just like making a file with a bunch of INCLUDEs.

I agree, I'd rather just keep a mask array around or use flags, but I am not going to tell @swiggumj that his preferred way of coding this is invalid and we refuse to support it on principle. I think TOAs objects having a unique object that identifies each TOA and what order it goes in is a useful thing, and until someone uses `merge_toas` the `index` does that. So I'm extending that to merged TOAs objects.

Reply:
A tuple column also has the same problem - the file numbers need to be bumped on merge as well.

Reply:
@scottransom Your proposal loses the property that `index` records position in the original file if any excision is done before merging. It also gives TOAs in the second and later files numbers that are larger than their position in those files. In fact, if no excision is done before merging, this does exactly the same thing as mine; if some excision has been done before merging, it renumbers all the TOAs after the first excision, while mine does not apart from files after the first. Excision itself does not renumber.

Reply:
Yes.  That is correct.  And I think I prefer it this way because `index` then always means the order of input TOAs that go into the newly created TOAs object.  It also means that `index` (in a newly created TOAs object) always has a min value that is 0 and a max value that is ntoas-1.  That seems much cleaner to me.  If you really want to track things from the original files (which we aren't even doing now with your proposal since commands etc aren't properly propagated) you can do that with a custom column.

As for your question above as to what I would propose for a custom column to track TOAs, I would personally just use `np.arange()` once I have the TOAs object that I want to start chopping up.  That's another reason why I think we should try to be consistent with the `index` column.

Reply:
The `index` column acquires holes as soon as excision is done. If no excision is done, my proposal agrees with yours. In both cases `index` indicates the order of TOAs.

Reply:
Right.  Excision is something that a user does.  So I'm fine with there being holes there.  What I don't like, and which I now think we included as a bug in `merge_TOAs`, is the fact that new TOAs objects `index` columns are not all created equally.  I think that they should be.

Reply:
I don't see a TOAs object that is the merge of excised TOAs objects as a "freshly created" TOAs object, I see it as "something that a user does" - just like the excision. This setup also provides a best-effort attempt at providing an understandable relationship between `index` values before and after the merge. If the user hasn't "done anything" before the merge, then the `index` column will be contiguous, as you suggest. If they have done excision, I prefer that it preserve the numbers as far as possible - which is exactly, for the first file, and the amount the second file is shifted by does not depend on what excision was done.

I think we are bikeshedding here, and I don't want it to delay this fix that our users need. The key invariant we need to establish is that TOAs objects do not contain repeated `index` values, and our two solutions agree on the simple case of loading a bunch of files and merging them before doing anything else. They also agree that when the user has done some excision the `index` column can have holes. They differ only on whether `merge_toas` "resets" the `index` column to be contiguous or preserves the numbers as well as possible. Can we just merge this?

Reply:
I wouldn't call this "bikeshedding" as we are talking about what is effectively an API change -- i.e. the meaning of `index`.  What you are proposing means that we cannot count on the `index` columns of merged TOAs objects having contiguous 0-(ntoas-1) indices.  What I'm recommending is that we fix the bug that we introduced so that that isn't the case.

I can see value in preserving an overall absolute TOA # in certain cases (which I think is what you are arguing for).  A user can still do that with additional columns.  I also see value in having the `index` column generated consistently every time we make a TOAs object.  And yes, a user could do this as well.

If we go the way that you suggest, we are baking in the choice that `index` does not have a set meaning anymore.  You have been the one who has argued very vocally and successfully about not changing our API without very good reason.  That's what I'm doing in this case!

Reply:
We differ on what we consider "making" a TOAs object. I consider loading TOAs from a file as something that should produce contiguous numbers (though I don't have my heart set on it). But merging, like subsetting, to me seems like manipulating existing TOAs objects. After all, using indexing to pull out a subset also constructs a new TOAs object, and I don't think either of us thinks that that should renumber all the TOAs. But for you merging does count as making a new TOAs object and thus we should renumber them; for me merging is one of a number of manipulations we can do on a TOAs object, and I would rather preserve the individual TOA indices as far as possible.

This is also something that can very easily be changed later.

Would it help to provide a `.renumber()` method for TOAs that remapped their numbers to be 0 to n-1, whether preserving the original order or using their `group_by` order? This could then be applied to the output of `merge_toas` or indexing for users who wanted contiguous numbers.

Reply:
I'd really like to get more opinions on this.  This change (and the bug we put in in `merge_TOAs` is *definitely* changing the intent as to what `index` was supposed to be (I know since I put it in there!).  I don't think we should change those things lightly (and I don't think you do either).

I think a `.renumber()` method would be fine, but it would simply be syntactic sugar (given that it is a one liner) *and* a user would have to know that they might want/need it!

I don't think it is cool that we are trying to ram something through that *definitely* breaks the (intended) API with calls that our users need it ASAP, when a lesser change (really a bugfix) could do just as well right now.

Reply:
Ok.  Please put a `.renumber()` method in, and I'll merge this.  That method would potentially help with the subset/indexing example you gave as well.

Reply:
Once we had TOAs objects that did not come from a single file (with INCLUDEs) we needed to establish what `index` should actually mean. I think it's perfectly fair to revisit this in the next few days and put in a PR if necessary to make the indices behave the way we decide is appropriate - perhaps worth discussing on the PINT telecon this week? In the meantime I think we should put in *something* to make index serve the purpose of unique identifiers. I'll write `.renumber()`.

Reply:
I think it is fine to talk about this on the PINT call (and after merging this).

Note that INCLUDE files are handled by the `index` column just fine (and exactly as the docstring says -- they are the order of the original individual TOAs when creating the TOAs object) because that is a serial process.  The issue really came up with `merge_TOAs` and the `__getitem__` returning new TOAs objects.  I totally agree we should have talked about it at those times.